sim/paladin/items.go

```diff
@@ -238,6 +240,62 @@ func init() {
 			},
 		})
 	})
+
+	core.NewItemEffect(LibramOfTheDevoted, func(agent core.Agent) {
+		paladin := agent.(PaladinAgent).GetPaladin()
+
+		actionID := core.ActionID{SpellID: 461309}
+		devotedMetrics := paladin.NewManaMetrics(actionID)
+
+		core.MakePermanent(paladin.RegisterAura(core.Aura{
+			Label: "Libram of the Devoted Trigger",
+			OnSpellHitTaken: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+				if result.DidBlock() {
+					amount := paladin.MaxMana() * 0.03
+					paladin.AddMana(sim, amount, devotedMetrics)
+				}
+			},
+		}))
+	})
+
+	core.NewItemEffect(LibramOfAvengement, func(agent core.Agent) {
+		paladin := agent.(PaladinAgent).GetPaladin()
+
+		paladin.OnSpellRegistered(func(spell *core.Spell) {
+			if spell.SpellCode == SpellCode_PaladinAvengersShield {
+				// Libram of Avenging causes Avenger's Shield to now hit the primary target
+				// twice (two projectiles).  Note: Avenger's Shield cannot miss or be resisted.
+				results := make([]*core.SpellResult, 2)
+				lowDamage := 366 * paladin.baseRuneAbilityDamage() / 100
+				highDamage := 448 * paladin.baseRuneAbilityDamage() / 100
+
+				spell.ApplyEffects = func(sim *core.Simulation, target *core.Unit, spell *core.Spell) {
+					apBonus := 0.091 * spell.MeleeAttackPower()
+					for idx := range results {
+						baseDamage := sim.Roll(lowDamage, highDamage) + apBonus
+						// Avenger's Shield cannot miss and uses magic critical _chance_.
+						results[idx] = spell.CalcDamage(sim, target, baseDamage, spell.OutcomeMagicCrit)
+					}
+					// The second projectile fires after a fixed 1.5s delay (2nd autocast does not trigger GCD)
+
+					// First hit
+					spell.WaitTravelTime(sim, func(sim *core.Simulation) {
+						spell.DealDamage(sim, results[0])
+					})
+
+					// Second hit
+					timeToHit := spell.TravelTime() + time.Millisecond*1500
+					core.StartDelayedAction(sim, core.DelayedActionOptions{
+						DoAt: sim.CurrentTime + timeToHit,
+						OnAction: func(s *core.Simulation) {
+							spell.DealDamage(sim, results[1])
+						},
+					})
+				}
+			}
```

Comment:
I wonder if this could just be a second cast of avenger's shield but then manually adjusting the cast metrics?

Reply:
Oh gotcha. Hm, I wonder if this should just be a toggle on the avenging wrath spell then with an attribute you can override here? I don't love the spell's code being copied here

Reply:
Yeah idk sometimes these effects are a total pain lol. You could just tie it to a paladin struct attribute and have the libram toggle that attribute on. Something like

`paladin.DoRighteousShieldSingleTarget` or something idk. In Wrath it was just a glyph IIRC